### ISSUE TITLE
NL.n  Use namespaces, do nest your namespaces. Do NOT use a suffixor or prefix in all defined names to provide scope.

### ISSUE THREAD

**Issue by SimonTemplate** [isocpp/CppCoreGuidelines#497]
Reason:
Provides a narrow context where defined names are short, external names are well marked and also naturally segments your header-files, this makes larger code sets more easy to browse and comprehend and aids with refactoring.

Chances are your IDE provides a folding mechanism for namespaces, which is great when browsing these larger pieces of code.

Examples:

Alternatives:

Exceptions:
extern “C”

Enforcement:
Lint defined names and suggest shortening of long names, find repeated prefixes.

See also:

Notes:
Do not follow the java path and start using DNS names, or any other global names.

Discussion:

**Comment by Pazns** [comment]
> Do not follow the java path and start using DNS names, or any other global names.

Can you be more clear :-) ?

**Comment by SimonTemplate** [comment]
@Pazns, I edited the line and tried to be clearer, hope it worked.

**Comment by LegalizeAdulthood** [comment]
This feels like many recommendations disguised as a single recommendation.  For guidance, I'd prefer to see a single recommendation for each point instead of lumped all together.  The "don't use DNS names" for organizing namespaces is sure to find some push-back particularly when coupled with the idea of using github to create yet-another-global-ad-hoc-registry for which the DNS already serves (which is why Java adopted that in the first place).

**Comment by SimonTemplate** [comment]
@LegalizeAdulthood well I agree, but as there are a number of subheaders in the rule template I got into adding more opinions, maybe I should split into a number of distinct recommendations.

**Comment by LegalizeAdulthood** [comment]
@SimonTemplate It's fine for it all to be lumped together as items discussed in the issue.  A pull request should split out the separate concerns.

**Comment by Pazns** [comment]
@SimonTemplate Yeah, thanks.

What is the idea behind this convention, by the way ?
What if a company doesn't have a domain name ?
How could namespaces benefit from a link with a domain name hierarchy like this ? Even if said namespaces don't exist ?

**Comment by SimonTemplate** [comment]
@Pazns Do NOT use reversed Internet domain names as a namespace hierarchy, is the idéa. I.e.there is no need for any domain names.

**Comment by Pazns** [comment]
@SimonTemplate Sorry it wasn't very understable, I was asking about this Java convention totally aside your proposition for the guidelines (with which I agree).

Is there any reasons for this DNS-style naming ? It seems a weird choice to me, bringing confusion of meaning if a real namespace hierarchy doesn't exist.

**Comment by SimonTemplate** [comment]
I suppose it's piggybacking on this DNS global naming authority, there are not many such authorities available. But the names contains too many levels and as Java also mirrors this into the folder structure it is rather unwieldy, IMHO.

**Comment by hpesoj** [comment]
I think trying to allocate top level namespaces would probably create more problems than it solved. If two popular libraries which are likely to be used together have conflicting namespaces, they can resolve it between themselves. And I can't imagine this happening all that often.

**Comment by SimonTemplate** [comment]
@hpesoj Well assuming a reasonable number of libraries, I suppose it will work.

**Comment by grahamreeds** [comment]
In Java (and .Net) it is natural to have nested namespaces and resulting
deep file hierarchies. It is only unwieldy in C++.  And even then you don't
have to nest your file structure.

In fact the only thing that makes it unwieldy is the inability to easily
nest namespaces.

Ie:
namespace foo { namespace bar { namespace stein {

As opposed to

namespace foo::bar::stein {

GR

**Comment by LegalizeAdulthood** [comment]
I thought the [N4230 proposal](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4230.html) was on track for C++17?

**Comment by grahamreeds** [comment]
But not currently standard.

Sorry for derailing the conversation slightly.

**Comment by LegalizeAdulthood** [comment]
@grahamreeds Agreed, not currently standard.

**Comment by SimonTemplate** [comment]
Yes, nested namespace definitions is welcome.

**Comment by AndrewPardoe** [comment]
This maybe shouldn't be its own guideline. We have a rule about header files--SF.7. Do we want to make a note about nesting as well as a note under this rule?

The rule about do not use suffixes or prefixes should be a note under the naming rules.

**Comment by SimonTemplate** [comment]
Yes and agree. This is not that major, more a convenience when digesting other peoples code.

**Comment by AndrewPardoe** [comment]
Assigning to Bjarne. To catch you up, we're looking at adding a note in SF.7 as the preferred option.

**Comment by SimonTemplate** [comment]
Great news. 

I am all for coding styles aimed at easing maintenance of large code sets, now only trying to promote this a bit more.

**Comment by johelegp** [comment]
Related: https://wg21.link/P0816 (No More Nested Namespaces in Library Design).

**Comment by MikeGitb** [comment]
@johelegp: Any Idea, how this was received by the committee? Apparently ranges are getting their own nested namespace instead of being put into std2.

**Comment by johelegp** [comment]
I don't remember having read on how they received it.
